README.md
dropClust: Efficient clustering of ultra-large scRNA-seq data
================


###   The 68K PBMC data analysis files are found in the `pbmc_68K_analysis` directory

   -   [Prerequisites](#prerequisites)
   -   [dropClust Tutorial](#dropclust-tutorial)
   -   [Filtering and Normalization](#filtering-and-normalization)
   -   [dropClust Sampling and Clustering](#dropclust-sampling-and-clustering)
   


Getting Started
===============

### Prerequisites:

1.  Python (&gt;=2.7), R
2.  Python `sklearn`, `igraph` packages

#### Linux and Mac Users
The C++ files will be build automatically when executing the main script for the first time.
The `igraph` python package may be installed using the command `pip install --user python-igraph`

#### For Windows users
The C++ Windows binaries are already provided, the user does not need to build them.
The `igraph` python package may be installed by downloading the appropriate pre-compiled binaries found at <https://www.lfd.uci.edu/~gohlke/pythonlibs/#python-igraph> followed by executing the command `python -m pip install desktop/path/igraph.whl`

For all platforms, the R dependencies will be installed automatically when executing the main script for the first time.


dropClust Tutorial
------------------

The goal of this tutorial is to guide the users to use dropClust. Most of the specific parameters have been explained so that users can adjust them specific to particular datasets or as per requirement. An analysis has been described in this tutorial to identify cell type specific genes. 



The demo data used for this tutorial has been obtained from <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65525> and is provided in a compresssed format (`es_mouse.zip`).



The provided data must be decompressed to obtain 3 files inside the `data/es_mouse` folder:

1.  matrix.mtx, a sparse matrix file containg FPKM values with genes in rows and samples in columns.
2.  barcodes.tsv, a single column file where each line represents a sampleID corresponding to the matrix columns.
3.  genes.tsv, a 2 column tab-delimited file where each line contains the (geneID, gene\_symbol) correspnding to the matrix rows.

Below is the corresponding Vigenette of the demo_dropClust_main.R file

#### Specify paths

``` r
sourceDir = getwd() # The directory hosting the current script 
setwd(sourceDir)
DATA_DIR <- file.path(sourceDir,"data/")        
FIG_DIR <-  paste0(sourceDir,"/plots/")        
REPORT_DIR  <- paste0(sourceDir,"/report/")   
LOUVAIN_DIR <- paste0(sourceDir,"/louvain/")  

dir.create(file.path(FIG_DIR),showWarnings = F)
dir.create(file.path(REPORT_DIR),showWarnings = F)
```

Load relevant libraries & functions for clustering

``` r
suppressMessages(source("libraries.R") )
suppressMessages(source("all_functions.R"))
```

Load data and annotations, the annotations may be omitted if unavailable.

``` r
mouse.data<-read10X(file.path(DATA_DIR,"es_mouse/"))
write.csv(x = mouse.data$gene_symbols, file = "gene_symbols.csv",quote = F,row.names =F)
annotation<- read.table(file.path(DATA_DIR,"annotations.csv"),sep = ',',header = T)
ref_id <- as.factor(annotation$x)
dim(mouse.data$mat)
```

    ## [1]  2717 24175

Filtering and Normalization
---------------------------

Filter poor quality cells. A threshold is provided such that the total UMI count across all genes of the filtered cells is greater than `th`. `th` corresponds to the total count of a cell.

``` r
filtered.data = filter_cells(mouse.data,th = 5000)
```

    ## [1] "470 bad cells present."

``` r
dim(filtered.data$mat)
```

    ## [1]  2247 24175

``` r
anno_labels= ref_id[filtered.data$keep_cells]
no_samples = dim(filtered.data$mat)[1]
```

Filter poor genes. Prior to normalization, the function `normalize_by_umi_2` filters genes which have UMI count greater than `min.count = 2` in atleast `min.cell = 3` cells. The dataset containing filtered cells and filtered genes is then UMI normalized.

``` r
lnorm<-normalize_by_umi_2(filtered.data, min.count=2, min.cell=3)   
```

    ## [1] "Dimensions of filtered Matrix:"
    ## [1] "2247 "  "21897 "

#### Matrix Subsetting

For minimising computation load, one third of the total samples is considered with a maximum of 20000 samples. Select Top Dispersed Genes as denoted by `ngenes_keep`. Write matrix subset to file. The genes are selected based on disperson statsstics.

``` r
set.seed(0)

i=min(20000, round(no_samples/3))
sample_ids = sample(1:no_samples, i)

m_n_ngenes <- matrix.subset(lnorm, sample_ids, ngenes_keep = 1000)
```

    ## [1] "Select variable Genes..."
    ## [1] "Sort Top Genes..."
    ## [1] "Cutoff Genes..."
    ## [1] "Writing Log Normalized whole_matrix, DIM: 2247 1000 ..."
    ## [1] "Writing Log Normalized sub_matrix, DIM: 749 1000 ..."

dropClust Sampling and Clustering
---------------------------------

#### Building neighbourhood graph using LSH.

LSH is performend on on `749` samples. This may take some time. Call python script to perform LSH based seach for nearest neighbours and estimate the neighbourhood graph. The default number of neighbours here is set to 10. The neighbourhood graph is partitioned by executing the louvain module in the shell. The present version of dropClust do not support any adjustment of the partition resolution. However the performance of dropClust does not depend on this as long as the scheme can distinguish the smaller group od population in the graph.

``` r
call_lsh()
```

    ## [1] "Reading sparce matrix..."            
    ## [2] ""                                    
    ## [3] "Converting matrix to dense format..."
    ## [4] "(749, 1000)"                         
    ## [5] "Initialize LSH..."                   
    ## [6] "Fit LSH..."                          
    ## [7] "Convert into adjacency matrix..."    
    ## [8] "Writing graph edgelist..."

``` r
call_louvain()
```

    ## [1] "Assiging Louvain Communities......................"

#### Sub-sampling

Read Louvain clusters for sub-sampling. Specify approximate number of samples to estimate the structure of the data. The samples are expected to sufficiently represent the entire data. The objective is to sample more form the smaller cluster. The `optimized_Pinit()` function uses simulated anneling to fine tune the exponential decay parameter to obtain the specified set of samples. To ensure selection of sufficient representative transcriptomes from small clusters, the exponential decay function used to determine the proportion of transciptomes to be sampled from each clusteris defined in the `sampling()` function. The functions had three floating parameters: `K, pfin and pinit`. The default values are set `K=500 and pfin=0.9` and `pinit` may be optimized to restrict the total number of sub-samples to the user defined value.

``` r
opt_pinit = optimized_Pinit(nsamples = 500) 
```

    ## Initializing par with random data inside bounds

``` r
# Sub-sampling using obtained parameter
subsamples_louvain<-sampling(pinit = opt_pinit)
write.csv(x = subsamples_louvain, file = "subsamples_idx",quote = F,row.names =F)
write.csv(x = filtered.data$barcodes[sample_ids[subsamples_louvain]], file = "barcodes_subsamples.csv",quote = F,row.names =F)

# Check the distribution of Sampling across annotated cell types. Skip if annotation is unavailable.
table(anno_labels[sample_ids[subsamples_louvain]])
```

    ## 
    ##  d0  d2  d4  d7 
    ## 194  59  69 178

The number of samples obtained at this satge `subsamples_louvain = 500` may not be an exact match with the input `nsamples` provided in the optimizer function.

#### Find PC Genes

Find PCA top 200 genes. This may take some time. 

``` r
top_pc_genes<-pc_genes(m_n_ngenes[subsamples_louvain,],top=200) 
```

![](README_files/figures/unnamed-chunk-9-1.png)

``` r
write.csv(x = top_pc_genes, file = "pc_gene_ids_sub.csv",quote = F,row.names =F)
```

#### Hierarchical Clustering on subsamples

Adjust Minimum cluster size with argument `minClusterSize` (default = 20)
Adjust tree cut with argument level `deepSplit` (default = 3)

``` r
ss_sel_genes_mat<-as.matrix(m_n_ngenes[subsamples_louvain,top_pc_genes])
ss_clusters<-ss_clustering(ss_sel_genes_mat, minClusterSize = 20, deepSplit = 3) 
```

    ## [1] "Predicted Clusters: 5"

#### Find K Nearest Neighbours among sub-samples.

The python script calls the LSHforest module to find the approximate K-nearest neighbous of the left put samples. The default value of K here is 5. The neighbours appearing here are among the cells samples previously.

``` r
system("python lsh/proj_neigh.py")
```

#### Un-annotated Cell Assignment

Class Assignment of Un-annotated cells i.e. the remaining cells which were left out due to sampling. For refwrence, clustering performance on the dropClust samples can be checked at this stage. The final clustering performance is guided by clusters at this stage.

``` r
INDEX  = read.csv("neigh.txt",header=F,sep=" ")
dim(INDEX)
```

    ## [1] 2247    5

``` r
clust_col<-cluster_assign(INDEX, ss_clusters)

write.csv(x = clust_col,file ="predicted.csv", quote = F,row.names = F)

sc_metric(clust_col, as.numeric(anno_labels),show_tab = F) # Skip if annotation is unavailable
```

    ##       ARI        RI    Purity 
    ## 0.8572084 0.9366013 0.9327993

### 2D Vizualization:

#### T-sne & Projection of co-ordinates of all samples in 2D

Compute Tsne Projection using PCA top genes on the dropClust samples followed by projecting TSNE co-ordinates on the remaining samples.

<span style="color:red"> **Alert!** The known *reference labels* is for technical demonstration purposes only. The biological significance of the predicted or reference clusters is not known. </span>

``` r
PROJ = compute_2d_embedding(data = as.matrix(m_n_ngenes[subsamples_louvain,top_pc_genes]), ss_clusters, INDEX)

dropClust_df<-as.data.frame(cbind(PROJ, clust_col))
rownames(dropClust_df)<-filtered.data$barcodes
save(dropClust_df,file="demo_proj.Rda")


# 2D Vizualization: Known References to Days.
plot_proj_df<-data.frame("Y1" = PROJ[,1],"Y2" = PROJ[,2],color =as.factor(anno_labels))
#plot_proj_df$color <- factor(plot_proj_df$color)
all_plot(plot_proj_df,"known.jpg","Known Day")
```

![](README_files/figures/known.jpg)

``` r
# 2D Vizualization: Predicted Clusters
plot_proj_df_pred<-data.frame(Y1 = PROJ[,1],Y2 = PROJ[,2],color = as.factor(clust_col))
#plot_proj_df_pred$color<-factor(plot_proj_df_pred$color)
all_plot(plot_proj_df_pred,"pred.jpg","Predicted clusters")
```

![](README_files/figures/pred.jpg)


NEWS.md

#### All changes to dropClust will be mentioned in this page.

Version 2.0.0 released
==

#### Change highlights
  - Available as a complete R package
  - Addressed memory bottle neck for faster performance; minimize RAM requirements.
  - Shinyapps Web server for dropClust [debsinha.shinyapps.io/dropClust/](https://debsinha.shinyapps.io/dropClust/)
  - A module dedicated for batch effect removal
  - Control parameter for post-hoc cluster assignment



#### The initial release has been moved to repository: [initial](https://github.com/debsin/dropClust-archives).


R/modules.R
#' Initial 1/3rd sampling if very big dataset
#' @param data_list list containing \code{read10x()} like structure
#' @param rare_data optional, return object obtained from the \code{get_rare_genes()} module.
#' @return two column integer matrix, first column represents sample id, second column contain cluster membership id
#' @export
initial.samples<-function(data_list,rare_data=NULL){

  if(is.null(rare_data)){
    rare_cells=c()
  }else{
    rare_cells= rare_data$rare_cells
  }
  # set.seed(0)

  no_samples = dim(data_list$mat)[1]

  i = ifelse(no_samples>20000, min(20000,round(no_samples/3)), no_samples)
  # i=min(20000, round(no_samples/3))
  sample_ids = sample(1:no_samples, i)

  sample_ids = union(sample_ids,which(data_list$barcodes %in% rare_cells))

  return(sample_ids)
}



#' ANN Graph Partition
#' @param data numeric matrix with \code{n} samples in rows
#' @return two column integer matrix, first column represents sample id, second column contain cluster membership id
#' @export
ann_partition<-function(data){
  dim(data)
  # set.seed(0)
  f = dim(data)[2]
  cat("Build graph...\n")
  t = methods::new(RcppAnnoy::AnnoyAngular,f)  # Length of item vector that will be indexed
  for(i in seq(nrow(data))){
    v = data[i,]
    t$addItem(i, v)
  }
  t$build(30)# 30 trees

  indices = lapply(seq(nrow(data)),function(x) t$getNNsByItem(x, 6))

  G=igraph::simplify(igraph::graph_from_adj_list(indices,
                                                 mode="all",
                                                 duplicate = FALSE))
  cat("Louvain Partition...\n")

  partition = igraph::cluster_louvain(G)

  dataMatrix =as.data.frame(cbind(seq(nrow(data)),partition$membership))
  return(dataMatrix)
}


#' ANN Search
#' @param data numeric matrix with \code{n} samples in rows
#' @param idx integers to specify specific sample indices. The ANN graph is fit on these samples only.
#' @param n_nn integers, specifies number of nearest neighbours.
#' @return integer matrix of dimension \code{n x n_nn}
#' @export
find_ann<-function(data,idx, n_nn = 10){
  n_nn = n_nn+1
  sub_mat = data[idx,]
  dim(sub_mat)
  cat("Build Graph...\n")
  f = dim(sub_mat)[2]
  t = methods::new(RcppAnnoy::AnnoyAngular,f)  # Length of item vector that will be indexed
  # set.seed(100)

  for(i in seq(nrow(sub_mat))){
    v = sub_mat[i,]
    t$addItem(i, v)
  }
  t$build(30) # 30 trees
  cat("Find ANN...\n")
  indices = mat.or.vec(nrow(data),n_nn)

  for(i in seq(nrow(data))){
    indices[i,] = t$getNNsByVector(data[i,], n_nn)
  }
  return(indices)
}






#' Cluster Validity Index
#' @description Validate clusters with known annotations when available.
#' @details Computes cluster external validity metrics in terms of Rand Index, Adjusted Rand Index and Purity. \cr
#' Purity of a cluster is measured as a recall of the largest known cell type within a predicted cluster. Total purity across all clusters is returned.
#' @param predicted_clusters vector of predicted cluster IDs.
#' @param known_types vector of annotated IDs corresponding to \code{predicted_clusters} in the same order.
#' @param show_table logical, When TRUE, shows the contingency table of concordance between the IDs.
#' @return numeric a vector of three metrics.
#' @export
sc_metric<-function(predicted_clusters ,known_types ,show_table = TRUE){
  dm<-table(known_types,predicted_clusters)
  if(show_table==TRUE){print(dm)}
  r.i. = flexclust::randIndex(dm, correct=TRUE, original=TRUE)
  return(round(c(r.i.,
                 "Purity" = ClusterPurity(predicted_clusters,
                                          known_types)), 3))
}

ClusterPurity <- function(clusters, classes) {
  sum(apply(table(clusters,classes), 2, max)) / length(clusters)
}



# ---------------------------------------------
# Optimize parameter 'pinit' for
# exponential decay sampling
# ---------------------------------------------
optimized_param<-function(partition, nsamples = 500){
  if(nsamples > nrow(partition)) stop("Expected number of samples exceeded total population.")
  oldseed = .Random.seed
  # set.seed(0)
  global.min <- 0
  tol <- 1e-3

  max.time= 20
  lower <- c(0.05, 0.9,500)
  upper <- c(0.1,0.95,4000)

  params<-NULL

  pin_find<-function(params, MAX_c = nsamples){
    output <- partition
    # oldseed = .Random.seed
    # set.seed(0)
    cluster_freq = table(output[,2])
    pinit = params[1]
    pfin = params[2]
    K=params[3]
    prop = round((pinit -
                    exp(-cluster_freq/K) * (pinit - pfin) )* cluster_freq)
    t(rbind(cluster_freq,prop))
    prop = reshape2::melt(prop)$value
    subsamples_louvain<-c()
    for(k in 1:length(prop)){
    subsamples_louvain = c(subsamples_louvain,
                             sample(output[which(output$V2==(k)),1],
                                    size = prop[k],replace = FALSE))
    }
    # .Random.seed = oldseed

    return(abs(MAX_c-length(subsamples_louvain)))
  }


  out <- GenSA::GenSA(par=params,fn = pin_find, lower = lower, upper = upper,
               control=list(max.time = max.time,
                            threshold.stop=global.min+tol,
                            verbose=TRUE))

  .Random.seed = oldseed

  #out[c("value","par","counts")]

  names(out$par) <-c("pinit", "pfin", "K")
  return(out$par)
}


# ---------------------------------------------
# Sampling Primary Clusters
# ---------------------------------------------
#' Sampling Primary Clusters
#' @description Performs sampling from the primary clusters in an inverse exponential order of cluster size.
#' @details Sampling in inverse proportion of cluster size following a exponential decay equation. To ensure selection of sufficient representative transcriptomes from small clusters, an exponential decay function  is used to determine the proportion of transcriptomes to be sampled from each cluster. For $i^{th}$ cluster, the proportion of expression profiles $p_i$ was obtained as follows.\cr
#' \ifelse{html}{\out{p<sub>i</sub> = p<sub>l</sub> <font face="symbol">-</font> e<sup><font face="symbol">-</font>(S<sub>i</sub>)/(K)</sup>}}{\deqn{Lp_{i} = p_{l} - e^{-\frac{S_i}{K}} (p_{l} - p_{u})}{ASCII}}
#' where \eqn{S_i} is the size of cluster \eqn{i}, \eqn{K} is a scaling factor, \eqn{p_i} is the proportion of cells to be sampled from the $i^{th}$ Louvain cluster. $p_l$ and $p_u$ are lower and upper bounds of the proportion value respectively.
#' @references {
#' \insertRef{sengupta2013reformulated}{dropClust}
#' }
#' @param data numeric matrix with \code{n} samples in rows
#' @param optm_parameters logical, when TRUE the parameters (\code{pinit, pfin, K}) are optimized such that exactly \code{nsamples} are returned. Optimization is performed using simulated annealing
#' @param nsamples integer, total number of samples to return post sampling; ignored when \code{optm_parameters = FALSE}.
#' @param pinit numeric [0,0.5], minimum probability of that sampling occurs from a cluster, ignored when \code{optm_parameters = TRUE}.
#' @param pfin numeric [0.5,1], maximum probability of that sampling occurs from a cluster, ignored when \code{optm_parameters = TRUE}.
#' @param K numeric, scaling factor analogous to Boltzman constant, ignored when \code{optm_parameters = TRUE}.
#' @return numeric vector containing the indices of sampled cells.
#' @export
sampling<-function(data, optm_parameters=FALSE,
                   nsamples=500,
                   pinit=0.195,
                   pfin = 0.9,
                   K=500){

  partition<-ann_partition(as.matrix(data))

  output<-partition
  if(optm_parameters==TRUE){
    param = optimized_param(partition, nsamples)
    pinit = param[1]
    pfin = param[2]
    K = param[3]
    cat("Optimized parameters:\n", param,"\n")
  }

  oldseed = .Random.seed
  # set.seed(0)
  cluster_freq = table(partition[,2])
  prop = round((pinit -
                  exp(-cluster_freq/K) * (pinit - pfin) )* cluster_freq)
  t(rbind(cluster_freq,prop))
  prop = reshape2::melt(prop)$value
  subsamples_louvain<-c()
  for(k in 1:length(prop)){
    subsamples_louvain = c(subsamples_louvain,
                           sample(partition[which(partition[,2]==(k)),1],
                                  size = prop[k],replace = FALSE))
  }
  .Random.seed = oldseed
  return(subsamples_louvain)
}


# ---------------------------------------------
# Select PCA based Genes
# ---------------------------------------------
#' Select PCA based Genes
#' @description Performs gene selection on sampled cells based on PCA loadings
#' @details Genes are ranked for selection in 3 steps:\cr
#' \enumerate{
#' \item First 50 principal components are obtained using Singular value Decomposition is used as implemented in the \code{irlba} R package.
#' \item Among the first 50 components, top 10 components are selected in the order of their modality.
#' \item Genes are ordered based on their average loadings in the rotation matrix containing the top 10 components.
#' }
#' @param mat normalized expression matrix containing the subset of samples. Subsetting the return matrix by \code{matrix.subset()} module with samples from primary clusters.
#' @param top integer specifying to number of genes to return in their order of ranking
#' @return integer vector containing the indices of \code{top} number of genes in the input matrix
#' @export
pc_genes<-function(mat,top=200){
  nc = 50
  cat("Find best PCA components...\n")
  PRR <- irlba::prcomp_irlba(as.matrix(mat),nc)
  S = c()
  for(i in 1:nc)
  {
    K = PRR$x[,i]
    # L = mclust::Mclust(K)
    # S = c(S,L$G)
    peaks = all_peaks(K)
    # modes= NbClust::NbClust(K, distance = "euclidean",
    #                min.nc = 2, max.nc = 10,
    #                method = "kmeans", index ="ball")$Best.nc[1]
    S = c(S,peaks)
    # cat(i,"\t")
  }
  graphics::barplot(S)

  df<-as.data.frame(cbind("MOD" = S,
                          "VAR" = apply(PRR$x[,1:nc],2, stats::var)))
  rownames(df)<-paste0("PC",seq(nc))
  #gene_pca <-abs(PRR$v[,which(S>=3)])
  best_components = utils::head(with(df, order(-MOD, -VAR)), 10)
  gene_pca <-abs(PRR$rotation[,best_components])
  gene_max<-apply(gene_pca, 1, max)
  rank_gene <- order(gene_max, decreasing=TRUE)


  top_pc_genes = utils::head(rank_gene,top)



  return(top_pc_genes)
}

all_peaks <- function(x) {
  den.fit <- stats::density(x, kernel=c("gaussian"))
  den.spline <- stats::smooth.spline(den.fit$x, den.fit$y,
                                     all.knots=TRUE, spar = 0.5)
  d1 <- stats::predict(den.spline, den.spline$x, deriv=1)
  peak.count <- length(rle(den.sign <- sign(d1$y))$values)/2
  if (is.na(peak.count) == TRUE) { peak.count <- 0 }
  return(peak.count)
}


# --------------------------------------------------
# 2D Projection
# --------------------------------------------------
#' Two Dimensional Embedding for Visualization
#' @description 2-D embedding of transcriptomes for visualization
#' @details Perms t-SNE for 2D embedding on sampled cells followed by the post-hoc projection of 2D co-ordinated of unsampled cells.
#' An ad-hoc clean up is also carried out to keep away stray points away from the plot - NA value is assigned to such points.
#' @param data normalized expression matrix containing the subset of samples and PCA based selected genes.
#' @param sp.samples transcriptome indices obtained from the Structure Preserving Sampling step.
#' @param clust.list list object as returned by \code{cluster.cells} module.
#' @return numeric two-column matrix containing the 2D embedding of all data points.
#' @export
compute_2d_embedding<-function(data, sp.samples, clust.list){
  # set.seed(0)
  # data = as.matrix(whole[sample_ids[subsamples_louvain[-ss_clusters$outliers]],top_pc_genes])
  # print("Compute Tsne Projection using PCA top genes...")
  ts<-Rtsne::Rtsne(as.matrix(data[sp.samples,]),
                   perplexity = 15,
                   dims = 2,
                   check_duplicates = FALSE)

  # print("Projecting TSNE co-ordinates...")
  tt<-ts$Y

  INDEX = clust.list$nn.ids
  clust_col = clust.list$cluster.ident
  clust_col[clust_col==0]<-NA

  sample.labels = clust_col[sp.samples]
  PROJ = matrix(NA, nrow=nrow(INDEX), ncol=2)

  nnn = ncol(INDEX)

  proj.idx = 1:nrow(INDEX)

  for( i in proj.idx)
  {
    if(is.na(clust_col[i])){next}
    maj_id = clust_col[i]
    maj_col = which(sample.labels[INDEX[i,]] == maj_id)
    # if(length(maj_col)/nnn > 0.5){
      # PROJ[i,] = apply(tt[INDEX[i,maj_col],],2,function(x) mean(x, na.rm = T))
      PROJ[i,1] = mean(tt[INDEX[i,maj_col],1], na.rm = TRUE)
      PROJ[i,2] = mean(tt[INDEX[i,maj_col],2], na.rm = TRUE)
    # }
  }


  # PROJ[sp.samples,] = tt

  return(PROJ)

}



# ---------------------------------------------
# Read  data
# ---------------------------------------------
#' Read data
#' @description Read 10X or csv data,
#' @details Loads expression data from 10X format directory or csv file as specified. The rows in the matrix are considered genes; the columns represent samples
#' @param data.path When format="10X", specify directory location containing 10X files: \cr
#' \enumerate{
#' \item matrix.mtx, sparse matrix format file
#' \item genes.tsv, gene names tab-separated file
#' \item barcodes.tsv, barcodes tab-separated file
#' }; with format="txt", specify the location of the csv matrix file.
#' @param format One of c("10X", "txt"). Defaults to "10X".
#' @param header TRUE/FALSE. Accept the header row to contain gene names within the csv matrix, defaults to TRUE. When FALSE, columns are labelled automatically.
#' @param sep mention the character separator, defaults to ','.
#' @param quote defaults to "\"".
#' @return list containing the 3 input files.
#' @importFrom utils read.csv
#' @export
read_data<-function(data.path, format="10X", header=TRUE, sep=',', quote="\""){




  data = switch(format,
                "10X" = read10X(data.path),
                "txt" = input_csv(data.path, header, sep, quote),
                stop('Invalid "format" specified, should be one of c("10X","txt")'))



  return(data)
}


#' @importFrom tibble has_rownames
input_csv<-function(x,  header, sep, quote){


  data = read.csv(x,
                  header = header,
                  sep = sep,
                  quote = quote)


  if(!(has_rownames(data))) {
    gene_symbols=paste("gene",formatC(1:nrow(data), width=floor(log10(nrow(data)))+1, flag="0"), sep="_")
  } else gene_symbols = rownames(data)


  if(header==T && class(data[,1])=="factor"){
    mat = data[,-1]
    gene_symbols = data[,1]
  }

  barcodes = colnames(mat)
  mat = as.matrix(mat)

  if(header==F) barcodes=paste("cell",formatC(1:ncol(mat), width=floor(log10(ncol(mat)))+1, flag="0"), sep="_")

  mat = t(mat)

  rownames(mat)<- barcodes
  colnames(mat)<-gene_symbols


  return(list("mat" = mat, "barcodes" = barcodes , "gene_symbols" = gene_symbols))
}


## reused Seurat code
read10X<-function(data.dir=NULL){
  full_data <- list()
  for(i in seq_along(data.dir)){
    run <- data.dir[i]
    if (!dir.exists(run)){
      stop("Directory provided does not exist")
    }

    if(!grepl("\\/$", run)){
      run <- paste(run, "/", sep = "")
    }

    barcode.loc <- paste(run, "barcodes.tsv", sep ="")
    gene.loc <- paste(run, "genes.tsv", sep ="")
    matrix.loc <- paste(run, "matrix.mtx", sep ="")

    if (!file.exists(barcode.loc)){
      stop("Barcode file missing")
    }
    if (!file.exists(gene.loc)){
      stop("Gene name file missing")
    }
    if (!file.exists(matrix.loc)){
      stop("Expression matrix file missing")
    }

    data <- Matrix::t(Matrix::readMM(matrix.loc))
    cell.names <- readLines(barcode.loc)
    gene.names <- readLines(gene.loc)
    if(all(grepl("\\-1$", cell.names)) == TRUE) {
      cell.names <- as.vector(as.character(sapply(cell.names,
                                                  extract_field,
                                                  1,
                                                  delim = "-")))
    }
    colnames(data) <- make.unique(as.character(sapply(gene.names,
                                                      extract_field,
                                                      2, delim = "\\t")))

    if(is.null(names(data.dir))){
      if(i < 2){
        rownames(data) <- cell.names
      }
      else {
        rownames(data) <- paste0(i, "_", cell.names, sep = "")
      }
    } else {
      rownames(data) <- paste0(names(data.dir)[i],"_",cell.names)
    }

  }

  full_data <- list("mat" = data,
                    "barcodes" = rownames(data) ,
                    "gene_symbols" = colnames(data))
  return(full_data)

}

extract_field=function(string,field=1,delim="_") {
  fields=as.numeric(unlist(strsplit(as.character(field),",")))
  if (length(fields)==1)  return(strsplit(string,delim)[[1]][field])
  return(paste(strsplit(string,delim)[[1]][fields],collapse = delim))
}
